File: cprdspy/CPR_matplotlib/Waves/waves.py
import matplotlib.pyplot as plt
import numpy as np
import sys
import os

from ..Arcs.arc import *
from ..Circles.circle import *

# from cprdspy.CPR_matplotlib.Arcs.arc import *

# # 方法 1: 使用相对导入（推荐用于包结构）
# from cprdspy.CPR_matplotlib.Circles.circle import *

# 方法 2: 使用 sys.path 绝对导入（适用于 Jupyter Notebook 或直接运行）
# 获取当前文件所在目录的父目录（CPR_matplotlib 目录）
# current_dir = os.path.dirname(os.path.abspath(__file__))
# parent_dir = os.path.dirname(current_dir)
# sys.path.insert(0, parent_dir)

# # 现在可以使用绝对导入
# from Circles.circle import *


def wave_ari(
    A=1,
    F=3,
    P=12,
    rotation=0,
    use_degree=True,
    direction="both",
    color="#0f0",
    alpha=1,
    R=1,
    center=(0, 0),
    points=1000,
    linestyle="-",
    linewidth=1,
    ax=None,
    show_center=True,
    **kwargs,
):
    """等差波形。

    rotation (float): 旋转角度（度或弧度取决于 use_degree）。
    use_degree (bool): 默认 True（度）。
    """
    rot_rad = np.deg2rad(rotation) if use_degree else rotation
    for i in range(P + 1):
        concentric_circles(
            center=(
                center[0] + np.cos(i * 2 * np.pi / P + np.pi / 2 + rot_rad),
                center[1] + np.sin(i * 2 * np.pi / P + np.pi / 2 + rot_rad),
            ),
            n=F,
            radius=R,
            param=A,
            mode="arithmetic",
            direction=direction,
            color=color,
            alpha=alpha,
            points=points,
            linestyle=linestyle,
            linewidth=linewidth,
            ax=ax,
            show_center=show_center,
            **kwargs,
        )


def wave_geo(
    A=1,
    F=3,
    P=12,
    rotation=0,
    use_degree=True,
    direction="both",
    color="#0f0",
    alpha=1,
    R=1,
    center=(0, 0),
    points=1000,
    linestyle="-",
    linewidth=1,
    ax=None,
    show_center=True,
    **kwargs,
):
    """等比波形。

    rotation (float): 旋转角度（度或弧度取决于 use_degree）。
    use_degree (bool): 默认 True（度）。
    """
    rot_rad = np.deg2rad(rotation) if use_degree else rotation
    for i in range(P + 1):
        concentric_circles(
            center=(
                center[0] + np.cos(i * 2 * np.pi / P + np.pi / 2 + rot_rad),
                center[1] + np.sin(i * 2 * np.pi / P + np.pi / 2 + rot_rad),
            ),
            n=F,
            radius=R,
            param=A,
            mode="geometric",
            direction=direction,
            color=color,
            alpha=alpha,
            points=points,
            linestyle=linestyle,
            linewidth=linewidth,
            ax=ax,
            show_center=show_center,
            **kwargs,
        )


def wave(
    A=1,
    F=3,
    P=12,
    rotation=0,
    use_degree=True,
    mode="geometric",
    direction="both",
    color="#0f0",
    alpha=1,
    R=1,
    center=(0, 0),
    points=1000,
    linestyle="-",
    linewidth=1,
    ax=None,
    show_center=True,
    **kwargs,
):
    """统一波形函数。

    rotation (float): 旋转角度（度或弧度取决于 use_degree）。
    use_degree (bool): 默认 True（度）。
    """
    rot_rad = np.deg2rad(rotation) if use_degree else rotation
    for i in range(P + 1):
        concentric_circles(
            center=(
                center[0] + np.cos(i * 2 * np.pi / P + np.pi / 2 + rot_rad),
                center[1] + np.sin(i * 2 * np.pi / P + np.pi / 2 + rot_rad),
            ),
            n=F,
            radius=R,
            param=A,
            mode=mode,
            direction=direction,
            color=color,
            alpha=alpha,
            points=points,
            linestyle=linestyle,
            linewidth=linewidth,
            ax=ax,
            show_center=show_center,
            **kwargs,
        )


def wave_wave(
    A=1,
    F=3,
    P=12,
    rotation=0,
    use_degree=True,
    mode="geometric",
    direction="both",
    color="#0f0",
    alpha=1,
    R=1,
    center=(0, 0),
    points=1000,
    linestyle="-",
    linewidth=1,
    ax=None,
    show_center=True,
    **kwargs,
):
    """简化波形函数。

    rotation (float): 旋转角度（度或弧度取决于 use_degree）。
    use_degree (bool): 默认 True（度）。
    """
    rot_rad = np.deg2rad(rotation) if use_degree else rotation
    for i in range(P + 1):
        wave(
            A=A, F=F, P=P,
            rotation=rot_rad, use_degree=False,
            mode=mode, direction=direction,
            color=color, alpha=alpha, R=R,
            center=(
                center[0] + np.cos(i * 2 * np.pi / P + np.pi / 2 + rot_rad),
                center[1] + np.sin(i * 2 * np.pi / P + np.pi / 2 + rot_rad),
            ),
            points=points, linestyle=linestyle, linewidth=linewidth,
            ax=ax, show_center=show_center, **kwargs,
        )

File: cprdspy/CPR_js1/Waves/waves.js
/**
 * CPR 图形库 - JavaScript 版本
 * Waves 模块 - 波形图案相关函数
 */

import { concentricCircles } from "../Circles/circle.js";

/**
 * 绘制等差波形
 * @param {Object} options - 配置选项
 * @param {number} options.A - 振幅（等差公差）
 * @param {number} options.F - 频率（圈数）
 * @param {number} options.P - 相位数
 * @param {number} options.rotation - 旋转角度
 * @param {string} options.direction - "both", "out", "in"
 * @param {string} options.color - 颜色
 * @param {number} options.alpha - 透明度
 * @param {number} options.R - 基准半径
 * @param {Array} options.center - 中心坐标
 * @param {number} options.points - 点数
 * @param {number} options.lineWidth - 线宽
 * @param {boolean} options.showCenter - 是否显示中心
 * @returns {Array} 所有圆形路径数据
 */
function waveAri({
    A = 1,
    F = 3,
    P = 12,
    rotation = 0,
    direction = "both",
    color = "#0f0",
    alpha = 1,
    R = 1,
    center = [0, 0],
    points = 1000,
    lineWidth = 1,
    showCenter = true,
} = {}) {
    const paths = [];

    for (let i = 0; i <= P; i++) {
        const circleCenter = [
            center[0] + Math.cos((i * 2 * Math.PI) / P + Math.PI / 2 + rotation),
            center[1] + Math.sin((i * 2 * Math.PI) / P + Math.PI / 2 + rotation),
        ];

        const circles = concentricCircles({
            center: circleCenter,
            n: F,
            radius: R,
            param: A,
            mode: "arithmetic",
            direction: direction,
            color: color,
            alpha: alpha,
            points: points,
            lineWidth: lineWidth,
            showCenter: showCenter,
        });

        paths.push(...circles);
    }

    return paths;
}

/**
 * 绘制等比波形
 * @param {Object} options - 配置选项
 * @param {number} options.A - 振幅（等比公比）
 * @param {number} options.F - 频率（圈数）
 * @param {number} options.P - 相位数
 * @param {number} options.rotation - 旋转角度
 * @param {string} options.direction - "both", "out", "in"
 * @param {string} options.color - 颜色
 * @param {number} options.alpha - 透明度
 * @param {number} options.R - 基准半径
 * @param {Array} options.center - 中心坐标
 * @param {number} options.points - 点数
 * @param {number} options.lineWidth - 线宽
 * @param {boolean} options.showCenter - 是否显示中心
 * @returns {Array} 所有圆形路径数据
 */
function waveGeo({
    A = 1,
    F = 3,
    P = 12,
    rotation = 0,
    direction = "both",
    color = "#0f0",
    alpha = 1,
    R = 1,
    center = [0, 0],
    points = 1000,
    lineWidth = 1,
    showCenter = true,
} = {}) {
    const paths = [];

    for (let i = 0; i <= P; i++) {
        const circleCenter = [
            center[0] + Math.cos((i * 2 * Math.PI) / P + Math.PI / 2 + rotation),
            center[1] + Math.sin((i * 2 * Math.PI) / P + Math.PI / 2 + rotation),
        ];

        const circles = concentricCircles({
            center: circleCenter,
            n: F,
            radius: R,
            param: A,
            mode: "geometric",
            direction: direction,
            color: color,
            alpha: alpha,
            points: points,
            lineWidth: lineWidth,
            showCenter: showCenter,
        });

        paths.push(...circles);
    }

    return paths;
}

/**
 * 绘制波形（通用）
 * @param {Object} options - 配置选项
 * @param {number} options.A - 振幅
 * @param {number} options.F - 频率
 * @param {number} options.P - 相位数
 * @param {number} options.rotation - 旋转角度
 * @param {string} options.mode - "arithmetic" 或 "geometric"
 * @param {string} options.direction - "both", "out", "in"
 * @param {string} options.color - 颜色
 * @param {number} options.alpha - 透明度
 * @param {number} options.R - 基准半径
 * @param {Array} options.center - 中心坐标
 * @param {number} options.points - 点数
 * @param {number} options.lineWidth - 线宽
 * @param {boolean} options.showCenter - 是否显示中心
 * @returns {Array} 所有圆形路径数据
 */
function wave({
    A = 1,
    F = 3,
    P = 12,
    rotation = 0,
    mode = "geometric",
    direction = "both",
    color = "#0f0",
    alpha = 1,
    R = 1,
    center = [0, 0],
    points = 1000,
    lineWidth = 1,
    showCenter = true,
} = {}) {
    const paths = [];

    for (let i = 0; i <= P; i++) {
        const circleCenter = [
            center[0] + Math.cos((i * 2 * Math.PI) / P + Math.PI / 2 + rotation),
            center[1] + Math.sin((i * 2 * Math.PI) / P + Math.PI / 2 + rotation),
        ];

        const circles = concentricCircles({
            center: circleCenter,
            n: F,
            radius: R,
            param: A,
            mode: mode,
            direction: direction,
            color: color,
            alpha: alpha,
            points: points,
            lineWidth: lineWidth,
            showCenter: showCenter,
        });

        paths.push(...circles);
    }

    return paths;
}

/**
 * 绘制嵌套波形
 * @param {Object} options - 配置选项
 * @param {number} options.A - 振幅
 * @param {number} options.F - 频率
 * @param {number} options.P - 相位数
 * @param {number} options.rotation - 旋转角度
 * @param {string} options.mode - "arithmetic" 或 "geometric"
 * @param {string} options.direction - "both", "out", "in"
 * @param {string} options.color - 颜色
 * @param {number} options.alpha - 透明度
 * @param {number} options.R - 基准半径
 * @param {Array} options.center - 中心坐标
 * @param {number} options.points - 点数
 * @param {number} options.lineWidth - 线宽
 * @param {boolean} options.showCenter - 是否显示中心
 * @returns {Array} 所有圆形路径数据
 */
function waveWave({
    A = 1,
    F = 3,
    P = 12,
    rotation = 0,
    mode = "geometric",
    direction = "both",
    color = "#0f0",
    alpha = 1,
    R = 1,
    center = [0, 0],
    points = 1000,
    lineWidth = 1,
    showCenter = true,
} = {}) {
    const paths = [];

    for (let i = 0; i <= P; i++) {
        const waveCenter = [
            center[0] + Math.cos((i * 2 * Math.PI) / P + Math.PI / 2 + rotation),
            center[1] + Math.sin((i * 2 * Math.PI) / P + Math.PI / 2 + rotation),
        ];

        const wavePaths = wave({
            A: A,
            F: F,
            P: P,
            rotation: rotation,
            mode: mode,
            direction: direction,
            color: color,
            alpha: alpha,
            R: R,
            center: waveCenter,
            points: points,
            lineWidth: lineWidth,
            showCenter: showCenter,
        });

        paths.push(...wavePaths);
    }

    return paths;
}

// 导出模块
export { waveAri, waveGeo, wave, waveWave };
